### PR TITLE
Widgets: make them compatible with WP 4.5 selective refresh

### PR DESCRIPTION
**Pull request by eliorivero**
This PR makes Jetpack widgets compatible with the new selective refresh feature introduced in Customizer in WP 4.5
It also fixes an issue introduced by #3543 where Facebook Like widget in selective refresh mode won't be displayed after making a change.

## Testing

Add
```php
add_theme_support( 'customize-selective-refresh-widgets' );
```
to the `functions.php` file in your theme.

### Notes

Introduce basic JS utilities to work with widgets in Customizer hooked to the wp object.

The function
```js
wp.customizerHasPartialWidgetRefresh();
```
checks that the Customizer supports partial widget refresh.
To use it and avoid breaking the front end when Customizer is not initialized, they must be used like:
```js
if ( wp && wp.customize && wp.customizerHasPartialWidgetRefresh() ) {
	// do something
}
```

The function
```js
wp.isJetpackWidgetPlaced( placement, widgetName );
```
verifies that the widget placed is the widget that dispatched the event. It requires the `placement` object returned by the event `partial-content-rendered` and the widget name defined in widget constructor.
For example, the gallery widget is defined as:
```php
parent::__construct(
	'gallery',
	/** This filter is documented in modules/widgets/facebook-likebox.php */
	apply_filters( 'jetpack_widget_name', __( 'Gallery', 'jetpack' ) ),
	$widget_ops,
	$control_ops
);
```
so the call is:
```js
wp.customize.selectiveRefresh.bind( 'partial-content-rendered', function( placement ) {
	if ( wp.isJetpackWidgetPlaced( placement, 'gallery' ) ) {
		tiledGalleries.findAndSetupNewGalleries();
	}
} );
```

### CODE REVIEW

File: modules/widgets/customizer-utils.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Utilities to work with widgets in Customizer.
+ */
+
+/**
+ * Checks whether this Customizer supports partial widget refresh.
+ * @returns {boolean}
+ */
+wp.customizerHasPartialWidgetRefresh = function() {
+	return 'object' === typeof wp && 'function' === typeof wp.customize && 'object' === typeof wp.customize.selectiveRefresh && 'object' === typeof wp.customize.widgetsPreview && 'function' === typeof wp.customize.widgetsPreview.WidgetPartial;
```
Comment (zinigor) [inline modules/widgets/customizer-utils.js]: Ouch, that's a long line. Nitpicking, but can we please have it split into several lines?

File: modules/widgets/customizer-utils.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+/**
+ * Verifies that the placed widget ID contains the widget name.
+ * @param {object} placement
+ * @param {string} widgetName
+ * @returns {*|boolean}
+ */
+wp.isJetpackWidgetPlaced = function( placement, widgetName ) {
+	var regex = new RegExp( '^' + widgetName + '-\\d+$', 'g' );
```
Comment (zinigor) [inline modules/widgets/customizer-utils.js]: Is it necessary to check for a digit after the widget name? Can we just use a string match and make sure that `widgetName` is strictly in the beginning?

**Comment by zinigor** [comment]
Tested all the widgets, everything works great, except for one notice I have fixed separately in 4bbec23, wanted to add it here, but accidentally committed and pushed it to `master`. Oops.

I have left a few comments, but everything seems great in general.

File: modules/widgets/customizer-utils.js (hunk 2 above)
Comment (eliorivero) [inline modules/widgets/customizer-utils.js]: I've kept the regex that Weston wrote. I think we can certainly check that only the widget name is in the string but it doesn't do any harm to check that it ends in a number as well.

**Comment by eliorivero** [comment]
Thanks for the fix, it's ok even if you pushed to master since it was something unrelated to this PR, thanks!

File: modules/widgets/customizer-utils.js (hunk 1 above)
Comment (eliorivero) [inline modules/widgets/customizer-utils.js]: Done :+1:

File: modules/widgets/customizer-utils.js (hunk 2 above)
Comment (eliorivero) [inline modules/widgets/customizer-utils.js]: On a second thought you do have a point, I've dismissed the usage of RegExp.

**Pull request merged**